### PR TITLE
tools refactor

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

File: langchain/tools/base.py [hunk 1]
```diff
@@ -123,3 +130,22 @@ async def arun(
...
             )
         return observation
+
+    def __call__(self, tool_input: str) -> str:
+        """Make tools callable with str input."""
+        return self.run(tool_input)
+
+    def run(self, tool_input: str, **kwargs: Any) -> str:
+        """Run the tool."""
+        if len(self.args) > 0:
```
Comment (vowelparrot) [inline langchain/tools/base.py:140]: Maybe I'm missing something but isn't `self.args` of len 1 by default?

Comment (hwchase17) [reply langchain/tools/base.py:140]: subclasses can override it

File: langchain/tools/base.py [hunk 2]
```diff
@@ -78,36 +92,43 @@ def run(
...
+        tool_input: Union[str, Dict],
         verbose: Optional[bool] = None,
         start_color: Optional[str] = "green",
         color: Optional[str] = "green",
         **kwargs: Any
     ) -> str:
         """Run the tool asynchronously."""
+        if isinstance(tool_input, str):
+            if len(self.args) > 1:
+                raise ValueError("Cannot call run on tools with > 1 argument")
```
Comment (nfcampos) [inline langchain/tools/base.py:104]: I think you might need to update this error message, something like "expected args of type dict, received of type string"

File: langchain/tools/base.py [hunk 3]
```diff
@@ -123,3 +144,7 @@ async def arun(
                 observation, verbose=verbose_, color=color, name=self.name, **kwargs
             )
         return observation
+
+    def __call__(self, tool_input: str) -> str:
+        """Make tools callable with str input."""
```
Comment (nfcampos) [inline langchain/tools/base.py:149]: This docstring isn't strictly true, as it will actually work to call this with a dict

File: langchain/tools/file_management/read.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from typing import List
+
+from langchain.tools.base import ArgInfo, BaseTool
+
+
+class ReadFileTool(BaseTool):
+    name: str = "read_file"
+    tool_args: List[ArgInfo] = [ArgInfo(name="file", description="name of file")]
```
Comment (nfcampos) [inline langchain/tools/file_management/read.py:8]: file_path might be a better name here

File: langchain/tools/python/tool.py [hunk 5]
```diff
@@ -28,13 +28,13 @@ class PythonREPLTool(BaseTool):
     python_repl: PythonREPL = Field(default_factory=_get_default_python_repl)
     sanitize_input: bool = True
 
-    def _run(self, query: str) -> str:
+    def _run(self, tool_input: str) -> str:
```
Comment (nfcampos) [inline langchain/tools/python/tool.py:31]: Might this be a breaking change for people who have written custom tools

**Review by nfcampos — COMMENTED** [review]
Some comments, lgtm otherwise

File: langchain/tools/python/tool.py (hunk 5 above)
Comment (hwchase17) [reply langchain/tools/python/tool.py:31]: good pt, i can look into making it less breaking

**Comment by vowelparrot** [comment]
Updated to use pydantic models for parsing and validation. Added a test to check that it works with args and kwargs specs. 

Need to update the decorator to work out of the box for structured functions. Will work on that now

**Comment by vowelparrot** [comment]
OK - updated the `Tool` class and decorator to permit structured args using the decorator.

**Comment by vowelparrot** [comment]
Added more unit tests and a bit to the `Tool` notebook. Will add another notebook on tool input validation

File: docs/modules/agents/tools/tool_input_validation.ipynb
Comment (agola11) [inline docs/modules/agents/tools/tool_input_validation.ipynb:1]: nit: having an exception thrown in a notebook with a big stack trace can clutter the docs that get built from this. Would suggest catching the exception and then doing somethign like `stack_trace = traceback.format_exc()`

File: langchain/tools/base.py [hunk 6]
```diff
@@ -35,39 +127,37 @@ def set_callback_manager(
...
         """Run the tool."""
+        run_input = self._parse_input(tool_input)
         if not self.verbose and verbose is not None:
             verbose_ = verbose
         else:
             verbose_ = self.verbose
         self.callback_manager.on_tool_start(
             {"name": self.name, "description": self.description},
-            tool_input,
+            str(run_input),
```
Comment (agola11) [inline langchain/tools/base.py:153]: nit: can just dump as json? run_input is a pydantic base model, so can just use `.json()`

File: tests/unit_tests/tools/requests/test_tool.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    RequestsDeleteTool,
+    RequestsGetTool,
+    RequestsPatchTool,
+    RequestsPostTool,
+    RequestsPutTool,
+    _parse_input,
+)
+
+
+class _MockTextRequestsWrapper(TextRequestsWrapper):
```
Comment (agola11) [inline tests/unit_tests/tools/requests/test_tool.py:17]: nit: This is not technically a mock, it's a Fake since you're stubbing the methods and not setting expectations

File: tests/unit_tests/tools/requests/test_tool.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    tool = RequestsGetTool(requests_wrapper=mock_requests_wrapper)
+    assert tool.run("https://example.com") == "get_response"
+    assert asyncio.run(tool.arun("https://example.com")) == "aget_response"
+
+
+def test_requests_post_tool(mock_requests_wrapper: TextRequestsWrapper) -> None:
+    tool = RequestsPostTool(requests_wrapper=mock_requests_wrapper)
+    input_text = '{"url": "https://example.com", "data": {"key": "value"}}'
+    assert tool.run(input_text) == "post {'key': 'value'}"
+    assert asyncio.run(tool.arun(input_text)) == "apost {'key': 'value'}"
```
Comment (agola11) [inline tests/unit_tests/tools/requests/test_tool.py:80]: probably no need to test the async methods here since you are really checking for the data passed in and they share the same code path. In the future, you can use `pytest.mark.asyncio` to await async code in tests without having to use asyncio.run() (which should really only be called once per application)

File: langchain/tools/base.py [hunk 9]
```diff
@@ -24,6 +89,33 @@ class Config:
...
+            # to be validated.
+            field_name = next(iter(pydantic_input_type.__fields__))
+            tool_input = {field_name: tool_input}
+        if pydantic_input_type is not None:
+            return pydantic_input_type.parse_obj(tool_input)
+        else:
+            raise ValueError(
+                f"args_schema required for tool {self.name} in order to"
+                f" accept input of type {type(tool_input)}"
+            )
```
Comment (agola11) [inline langchain/tools/base.py:117]: we can't hit this right? Because if the `pydantic_input_type` is not defined (i.e. self.args), we `create_args_schema_model_from_signature`?

File: langchain/tools/base.py [hunk 10]
```diff
@@ -1,19 +1,84 @@
...
 
 from langchain.callbacks import get_callback_manager
 from langchain.callbacks.base import BaseCallbackManager
 
 
-class BaseTool(BaseModel):
-    """Class responsible for defining a tool or skill for an LLM."""
+def create_args_schema_model_from_signature(run_func: Callable) -> Type[BaseModel]:
+    """Create a pydantic model type from a function's signature."""
+    signature_ = inspect.signature(run_func)
```
Comment (agola11) [inline langchain/tools/base.py:15]: This code is a bit hard to follow for me lol, do we really need to do this? Why can't we just just enforce that tools which take in `dict` params moving forward have to define their input schema as pydantic? Seems simpler than having to create a model on the fly
